src/plugins/oidc/login.ts

```diff
@@ -57,70 +71,106 @@ const authorize = async (org: OrgData, scope: string) => {
  * authorization server will return "authorization_pending", in which case this
  * function will return undefined.
  */
-const fetchOidcToken = async (org: OrgData, authorize: AuthorizeResponse) => {
-  if (org.providerType === undefined) {
-    throw "Login requires a configured provider type.";
-  }
-  const init = {
-    method: "POST",
-    headers: OIDC_HEADERS,
-    body: urlEncode({
-      client_id: org.clientId,
-      device_code: authorize.device_code,
-      grant_type: DEVICE_GRANT_TYPE,
-    }),
-  };
-  validateProviderDomain(org);
-  const url =
-    org.providerType === "okta"
-      ? `https:${org.providerDomain}/oauth2/v1/token`
-      : org.providerType === "ping"
-        ? `https://${org.providerDomain}/${org.environmentId}/as/token`
-        : throwAssertNever(org.providerType);
+export const fetchOidcToken = async <T>(request: {
+  url: string;
+  init: RequestInit;
+}) => {
+  const { url, init } = request;
   const response = await fetch(url, init);
-
   if (!response.ok) {
     if (response.status === 400) {
       const data = await response.json();
       if (data.error === "authorization_pending") return undefined;
     }
     await validateResponse(response);
   }
-  return (await response.json()) as TokenResponse;
+  return (await response.json()) as T;
+};
+
+const oidcRequestBuilder = (org: OrgData, authorize: AuthorizeResponse) => {
+  if (org.providerType === undefined) {
+    throw "Login requires a configured provider type.";
+  }
+  validateProviderDomain(org);
+
+  return {
+    url:
+      org.providerType === "okta"
+        ? `https:${org.providerDomain}/oauth2/v1/token`
+        : org.providerType === "ping"
+          ? `https://${org.providerDomain}/${org.environmentId}/as/token`
+          : throwAssertNever(org.providerType),
+    init: {
+      method: "POST",
+      headers: OIDC_HEADERS,
+      body: urlEncode({
+        client_id: org.clientId,
+        device_code: authorize.device_code,
+        grant_type: DEVICE_GRANT_TYPE,
+      }),
+    },
+  };
 };
 
 /** Waits until user device authorization is complete
  *
  * Returns the OIDC token after completion.
  */
-const waitForActivation = async (
-  org: OrgData,
-  authorize: AuthorizeResponse
-) => {
-  const start = Date.now();
-  while (Date.now() - start <= authorize.expires_in * 1e3) {
-    const response = await fetchOidcToken(org, authorize);
-    if (!response) await sleep(authorize.interval * 1e3);
-    else return response;
-  }
-  throw "Expired awaiting in-browser authorization.";
-};
+export const waitForActivation =
+  <A, T>(authorize: A & { interval: number }) =>
+  async (
+    getExpiry: (authorize: A) => number, // Aws implementation differs from standard OIDC response, need function to extract expiry
```

Comment:
Clarity on this comment: does this mean that the AWS Identity Center is different than Okta SAML?

Reply:
aws oidc implementation does not conform to the standard oidc spec. especially the naming convention.